lib/jsdom/living/helpers/geometry/matrixDomStringParser.js

```diff
@@ -0,0 +1,397 @@
+  grad(x) {
+    return this.turn(x) / 400;
+  },
+  rad(x) {
+    return x * (180 / Math.PI);
+  },
+  turn(x) {
+    return x * 360;
+  }
+};
+
+const paramConverter = {
+  _rules: {
+    number(x) {
+      return (SYMBOL.radixPoint.test(x) || !isNaN(x)) && x !== null;
+    },
+    angleUnit(x) {
+      return angleUnits.includes(x);
+    },
+    absoluteLengthUnit(x) {
+      return absoluteLengths.includes(x);
+    }
+  },
+  _validate(rules, param) {
+    rules.forEach(matcher => {
+      if (!matcher(param)) {
+        throw new TypeError(`Failed to validate a parameter, ${rules} ${param}`);
+      }
+    });
+  },
+  _getUnitParam(paramString) {
+    let quantity = "";
+    let unit = "";
+
+    if (typeof paramString !== "string") {
+      throw new TypeError();
+    }
+
+    for (const next of paramString) {
+      if (this._rules.number(next) && unit.length === 0) {
+        quantity += next;
+      } else {
+        unit += next;
+      }
+    }
+
+    if (
+      quantity.length === 0
+    ) {
+      throw new TypeError();
+    }
+
+    return [quantity, unit];
+  },
+  number(param) {
+    const { number } = this._rules;
+    this._validate([number], param);
+
+    return Number(param);
+  },
+  lengthPercentage(param) {
+    const { number, absoluteLengthUnit } = this._rules;
+    const [length, unit] = this._getUnitParam(param);
+    this._validate([number], length);
+    this._validate([absoluteLengthUnit], unit);
+
+    return LENGTH_CONVERTERS[unit](Number(length));
+  },
+  angle(param) {
+    const { number, angleUnit } = this._rules;
+    const [length, unit] = this._getUnitParam(param);
+
+    this._validate([number], length);
+    // All transform functions expect 0 or angle type
+    if (Number(length) !== 0) {
+      this._validate([angleUnit], unit);
+    }
+
+    return ANGLE_CONVERTERS[unit || "deg"](Number(length));
+  }
+};
+
+// https://drafts.csswg.org/css-transforms-1/#two-d-transform-functions
+// https://drafts.csswg.org/css-transforms-2/#three-d-transform-functions
+
+// Contains the definitions for valid parameters per transform function.
+// Preforms conversion of parameters to expected values.
+const transformFnParameterValidation = {
+  matrix3d: (...params) => {
+    validateParameterLength(16, params, "matrix3d");
+    return params.map(paramConverter.number.bind(paramConverter));
+  },
+  matrix: (...params) => {
+    validateParameterLength(6, params, "matrix");
+    return params.map(paramConverter.number.bind(paramConverter));
+  },
+  translate3d: (...params) => {
+    validateParameterLength(3, params, "translate3d");
+    return [
+      paramConverter.lengthPercentage(params[0]),
+      paramConverter.lengthPercentage(params[1]),
+      paramConverter.lengthPercentage(params[2])
+    ];
+  },
+  translate: (...params) => {
+    validateParameterLength([1, 2], params, "translate");
+    return [
+      paramConverter.lengthPercentage(params[0]),
+      paramConverter.lengthPercentage(params[1])
```

Comment:
By spec, this should default to zero if `params[1] === undefined`. However, it doesn't _look_ like lengthPercentage handles undefined(?)